scene_graph/vrd_tutorial.py

```diff
@@ -0,0 +1,335 @@
+
+# %%
+# %load_ext autoreload
+# %autoreload 2
+
+import numpy as np
+
+# %% [markdown]
+# ### 1. Load Dataset
+# We load the VRD dataset and filter images with at least one semantic predicate in it. We load the train, valid, and test sets as Pandas DataFrame objects with the following fields:
+# - label: The relationship between the objects. 0: `RIDE`, 1: `CARRY`, 2: `OTHER` action predicates
+# - object_bbox: coordinates of the bounding box for the object `[ymin, ymax, xmin, xmax]`
+# - object_category: category of the object
+# - source_img: filename for image the relationship is in
+# - subject_bbox: coordinates of the bounding box for the object `[ymin, ymax, xmin, xmax]`
+# - subject_category: category of the subject
+#
+# Note that the `train_df` object has a labels field with all -1s. This denotes the lack of labels for that particular dataset. In this tutorial, we will assign probabilistic labels to the training set by writing labeling functions over attributes of the subject and objects!
+
+# %%
+import os
+if os.path.basename(os.getcwd()) == "scene_graph":
+    os.chdir("../")
+
+# %%
+from scene_graph.utils import load_vrd_data
+
+train_df, valid_df, test_df = load_vrd_data()
+
+print("Train Relationships: ", len(train_df))
+print("Dev Relationships: ", len(valid_df))
+print("Test Relationships: ", len(test_df))
+
+# %% [markdown]
+# ## 2. Writing Labeling Functions
+# We now write labeling functions to detect what relationship exists between pairs of bounding boxes. To do so, we can encode knowledge about the categories of subjects and objects usually involved in these relationships (e.g., `person` is usually the subject for predicates like `ride` and `carry`). We can encode common knowledge about these predicates, such as the subject is usually higher than the object for the predicate `ride` into the labeling functions.
+
+# %%
+RIDE = 0
+CARRY = 1
+OTHER = 2
+ABSTAIN = -1
+
+
+from snorkel.labeling.lf import labeling_function
+# Category-based LFs
+@labeling_function()
+def LF_ride_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bike", "snowboard", "motorcycle", "horse"]:
+            return RIDE
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_ride_rare_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bus", "truck", "elephant"]:
+            return RIDE
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_carry_object(x):
+    if x.subject_category == "person":
+        if x.object_category in ["bag", "surfboard", "skis"]:
+            return CARRY
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_carry_subject(x):
+    if x.object_category == "person":
+        if x.subject_category in ["chair", "bike", "snowboard", "motorcycle", "horse"]:
+            return CARRY
+    return ABSTAIN
+
+
+@labeling_function()
+def LF_person(x):
+    if x.subject_category != "person":
+        return OTHER
+    return ABSTAIN
+
+
+# Distance-based LFs
```

Comment:
let's separate each set of LFs and write a sentence describing them?

Reply:
I added paragraphs above each set instead of each LF, they're not too complex